File: frontend/src/types/changes.ts
export interface ChangeLine {
  id: string
  cost_account_id: string
  cost_account_code: string
  cost_account_description: string
  hour_impact: number
  cost_impact: number
}

export interface ChangeOrder {
  id: string
  change_code: string
  description: string | null
  category: 'budget_transfer' | 'scope' | 'growth' | 'trend' | null
  status: 'pending' | 'submitted' | 'approved' | 'cancelled' | string
  reason: string | null
  impact: 'cost' | 'schedule' | 'both' | 'none'
  request_date: string | null
  issued_date: string | null
  approved_date: string | null
  scope_notes: string | null
  added_days: number | null
  pct_complete: number | null
  total_hour_impact: number
  total_cost_impact: number
}

export interface ChangeOrderDetail extends ChangeOrder {
  lines: ChangeLine[]
}

export interface ChangeOrderIn {
  change_code: string
  description?: string
  category?: string
  status?: string
  reason?: string
  impact?: string
  request_date?: string | null
  issued_date?: string | null
  approved_date?: string | null
  scope_notes?: string
  added_days?: number | null
  pct_complete?: number | null
}

export interface WbsChangeItem {
  change_code: string
  description: string | null
  status: string
  total_cost_impact: number
  total_hour_impact: number
  request_date: string | null
}

export interface ChangeOrderUpdate {
  description?: string
  category?: string | null
  status?: string
  reason?: string | null
  impact?: string
  request_date?: string | null
  issued_date?: string | null
  approved_date?: string | null
  scope_notes?: string | null
  added_days?: number | null
  pct_complete?: number | null
}

File: backend/app/schemas/changes.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class ChangeLineOut(BaseModel):
    id: UUID
    cost_account_id: UUID
    cost_account_code: str
    cost_account_description: str
    hour_impact: float
    cost_impact: float
    qty_element_id: UUID | None = None
    qty_element_code: str | None = None
    qty_element_unit: str | None = None
    qty_scope_impact: float | None = None


class ChangeOrderOut(BaseModel):
    id: UUID
    change_code: str
    description: str | None
    category: str | None
    status: str
    reason: str | None
    impact: str
    request_date: datetime | None
    issued_date: datetime | None
    approved_date: datetime | None
    scope_notes: str | None
    added_days: float | None
    pct_complete: float | None
    total_hour_impact: float
    total_cost_impact: float


class ChangeOrderDetailOut(ChangeOrderOut):
    lines: list[ChangeLineOut]


class ChangeOrderIn(BaseModel):
    change_code: str
    description: str | None = None
    category: str | None = None
    status: str = "pending"
    reason: str | None = None
    impact: str = "cost"
    request_date: datetime | None = None
    issued_date: datetime | None = None
    approved_date: datetime | None = None
    scope_notes: str | None = None
    added_days: float | None = None
    pct_complete: float | None = None


class WbsChangeItemOut(BaseModel):
    change_code: str
    description: str | None
    status: str
    total_cost_impact: float
    total_hour_impact: float
    request_date: datetime | None


class ChangeLineIn(BaseModel):
    account_code: str
    hour_impact: float = 0.0
    cost_impact: float = 0.0
    qty_element_code: str | None = None
    qty_scope_impact: float | None = None


class ChangeLineUpdate(BaseModel):
    hour_impact: float | None = None
    cost_impact: float | None = None
    qty_element_code: str | None = None
    qty_scope_impact: float | None = None


class ChangeOrderUpdate(BaseModel):
    description: str | None = None
    category: str | None = None
    status: str | None = None
    reason: str | None = None
    impact: str | None = None
    request_date: datetime | None = None
    issued_date: datetime | None = None
    approved_date: datetime | None = None
    scope_notes: str | None = None
    added_days: float | None = None
    pct_complete: float | None = None
